packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/serializer/BsonEncoderTests.kt

```diff
@@ -0,0 +1,371 @@
+                bsonValue = BsonString("")
+            )
+        }
+    }
+
+    private class DecoderAsserter(
+        val deserializedType: KClass<*>,
+        bsonType: KClass<*>? = null,
+        val bsonValue: BsonValue,
+        val value: Any? = null
+    ) {
+        val bsonType: KClass<*> = bsonType ?: bsonValue::class
+    }
+
+    private val wrongPrimitiveAsserters = listOf(
+        DecoderAsserter(
+            deserializedType = Byte::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Short::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Int::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Long::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Float::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Double::class,
+            bsonType = BsonNumber::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = Boolean::class,
+            bsonType = BsonBoolean::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = String::class,
+            bsonType = BsonString::class,
+            bsonValue = BsonInt32(0)
+        ),
+        DecoderAsserter(
+            deserializedType = Char::class,
+            bsonType = BsonString::class,
+            bsonValue = BsonInt32(0)
+        ),
+        DecoderAsserter(
+            deserializedType = ByteArray::class,
+            bsonType = BsonBinary::class,
+            bsonValue = BsonInt32(0)
+        ),
+    )
+
+    private val wrongRealmValueAsserters = listOf(
+        DecoderAsserter(
+            deserializedType = RealmUUID::class,
+            bsonType = BsonBinary::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = MutableRealmInt::class,
+            bsonType = BsonInt64::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = ObjectId::class,
+            bsonType = BsonObjectId::class,
+            bsonValue = BsonString("")
+        ),
+        DecoderAsserter(
+            deserializedType = RealmInstant::class,
+            bsonType = BsonDateTime::class,
+            bsonValue = BsonString("")
+        ),
+    )
+
+    @Test
+    fun decodeFromBsonElement_throwsWrongType() {
+        (wrongPrimitiveAsserters + wrongRealmValueAsserters).forEach {
```

Comment:
Nothing serious, but since we don't really uses these variables elsewhere we could maybe just define them locally in the method. Also a bit hard to track because you need to look up the `DecoderAsserter` class to understand it. Why not just use a `Triple` and have all the logic here and then destruct them into named variables to that it was easier to follow. Anyway, nothing critical 🤷

Reply:
The asserters data structure has 4 parameters and is being reused in several tests. 

Yes, we could use a `Triple` because not all parameters are used in every test, but it would mean that we would have to define multiple similar datasets in several tests..